elastic/apm-server
# benchmark: Introduce new binary `intake-receiver`

## Motivation/summary

Adds a new binary under `cmd/intake-receiver` which is a simplified
implementation of the APM Server's intake API. Instead of persisting the
events to Elasticsearch, the events are stored in _ndjson_ files. Each
of these files the contains traces received by the APM Agents, keeping
the batches just as they are received to be able to replay them during
the benchmark execution phase. The format for these files is:

    traces/<agent_name>-<agent_version>.ndjson

This allows us to classify each version of an APM Agent individually and
possibly benchmark and compare different versions of same application in
the future.

The `intake-receiver` doesn't perform aggregations or produce metrics
from the received traces, it simply stores them in plain-text.

For the PoC phase the `intake-receiver` will be used to store traces as
they are generated by the opbeans apps in the `apm-integration-testing`
repository.

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~

## How to test these changes

1. `cd cmd/intake-receiver && go build .`.
2. `./intake-receiver`
3. `curl -H "Content-type: application/x-ndjson" --data-binary @testdata/intake-v3/rum_events.ndjson http://localhost:8200/intake/v3/rum/events`
4. `curl -H "Content-type: application/x-ndjson" --data-binary @testdata/intake-v2/transactions.ndjson http://localhost:8200/intake/v2/events`
5. `curl localhost:8200/` 
```
{"publish_ready":true,"version":"8.2.0"}
```
6. Check there are logs in the intake receiver process such as:
```
2022/02/24 21:26:24 processed request to /intake/v3/rum/events (curl/7.77.0) in 790.5µs
2022/02/24 21:26:35 processed request to /intake/v2/events (curl/7.77.0) in 467.084µs
2022/02/24 21:27:12 processed request to / (curl/7.77.0) in 8.416µs
```
7. Close the intake receiver with an interrupt (ctrl+c) and verify it closes the files it has open:
```
^C2022/02/24 21:28:02 closing http server...
2022/02/24 21:28:02 closed file traces/js-base-4.8.1.ndjson
2022/02/24 21:28:02 closed file traces/elastic-node-3.14.0.ndjson
```
8. Verify that stored files match the source files that were used on the test requests.

## Related issues

Part of #7216
